gprofiler/spark/spark_collector.py

```diff
@@ -0,0 +1,894 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+# (C) Datadog, Inc. 2018-present. All rights reserved.
+# Licensed under a 3-clause BSD style license (see LICENSE.bsd3).
+#
+import json
+import os
+import re
+import time
+import traceback
+import xml.etree.ElementTree as ET
+from datetime import datetime
+from threading import Event, Lock, Thread
+from typing import Any, Dict, Generator, List, Optional, Tuple, Union, cast
+from urllib.parse import urljoin, urlparse
+
+import psutil
+import requests
+from requests.exceptions import ConnectionError, HTTPError, InvalidURL, Timeout
+
+from gprofiler.client import APIClient
+from gprofiler.log import get_logger_adapter
+from gprofiler.metadata.system_metadata import get_hostname
+from gprofiler.platform import is_windows
+from gprofiler.spark.metrics import (
+    ACTIVE_EXECUTORS_COUNT,
+    EXECUTORS_COUNT,
+    SPARK_APPLICATION_DIFF_METRICS,
+    SPARK_APPLICATION_GAUGE_METRICS,
+    SPARK_RDD_METRICS,
+    SPARK_STAGE_METRICS,
+    SPARK_STREAMING_BATCHES_METRICS,
+    SPARK_STREAMING_STATISTICS_METRICS,
+    SPARK_STRUCTURED_STREAMING_METRICS,
+    SPARK_TASK_SUMMARY_METRICS,
+    YARN_CLUSTER_METRICS,
+    YARN_NODES_METRICS,
+)
+from gprofiler.utils import get_iso8601_format_time
+from gprofiler.utils.process import search_for_process
+
+GMT_TIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%fGMT"
+
+# Application type and states to collect
+YARN_SPARK_APPLICATION_SPECIFIER = "SPARK"
+YARN_RUNNING_APPLICATION_SPECIFIER = "RUNNING"
+
+# SPARK modes
+SPARK_DRIVER_MODE = "driver"
+SPARK_YARN_MODE = "yarn"
+SPARK_MESOS_MODE = "mesos"
+
+# COMMON urls
+YARN_APPS_PATH = "ws/v1/cluster/apps"
+YARN_CLUSTER_PATH = "ws/v1/cluster/metrics"
+YARN_NODES_PATH = "ws/v1/cluster/nodes"
+SPARK_APPS_PATH = "api/v1/applications"
+MESOS_MASTER_APP_PATH = "/frameworks"
+
+# METRIC keys
+METRIC_TIMESTAMP_KEY = "timestamp"
+METRICS_DATA_KEY = "metrics"
+METRICS_FILE_PREFIX = "spark_metric_"
+
+STRUCTURED_STREAMS_METRICS_REGEX = re.compile(
+    r"^[\w-]+\.driver\.spark\.streaming\.(?P<query_name>[\w-]+)\.(?P<metric_name>[\w-]+)$"
+)
+
+
+class SparkCollector:
+    def __init__(
+        self,
+        cluster_mode: str,
+        master_address: str,
+        cluster_metrics: bool = True,
+        applications_metrics: bool = True,
+        streaming_metrics: bool = True,
+    ) -> None:
+        self._lock = Lock()
+        self._logger = get_logger_adapter(__name__)
```

Comment:
Can be coherent with the rest of the codebase and set up `logger` as a global object in this module, instead of `self._logger`.